hxnormalize.1
.TH "HXNORMALIZE" "1" "10 Jul 2011" "7.x" "HTML-XML-utils"
.SH NAME
hxnormalize \- pretty-print an HTML file
.SH SYNOPSIS
.B hxnormalize
.RB "[\| " \-x " \|]"
.RB "[\| " \-X " \|]"
.RB "[\| " \-e " \|]"
.RB "[\| " \-d " \|]"
.RB "[\| " \-s " \|]"
.RB "[\| " \-L " \|]"
.RB "[\| " \-i
.IR indent " \|]"
.RB "[\| " \-l
.IR line\-length " \|]"
.RB "[\| " \-c
.IR commentmagic " \|]"
.RI "[\| " file-or-URL " \|]"
.SH DESCRIPTION
.LP
The
.B hxnormalize
command pretty-prints an HTML or XML file, and also tries to fix small
HTML errors. The output is the same file, but with a maximum line length
and with optional indentation to indicate the nesting level of each
line.
.SH OPTIONS
The following options are supported:
.TP 10
.B \-x
Applies XML conventions: empty elements are written with a slash at
the end (e.g., <IMG\ />) and, if the input is HTML, any \(oq<\(cq and
\(oq&\(cq inside <style> and <script> elements are escaped as
\(oq&lt;\(cq and \(oq&amp;\(cq. (The input is assumed to be HTML
unless the
.B \-X
option is present.) Implies
.BR \-e .
.TP
.B \-e
Always inserts endtags, even if HTML does not require them (for
example: </p> and </li>).
.TP
.B \-X
Makes
.B hxnormalize
assume the input is well-formed XML. It does not try to infer omitted
HTML tags, does not assume elements such as <img> and <br> are empty,
and does not treat \(oq<\(cq and \(oq&\(cq inside <style> and <script>
as normal characters.
.TP
.B \-d
Omit the DOCTYPE from the output.
.TP
.BI \-i " indent"
Set the number of spaces to indent each nesting level. Default is 2.
Not all elements cause an indent. In general, elements that can occur
in a block environment are started on a new line and cause an indent,
but inline elements, such as EM and SPAN do not cause an indent.
.TP
.BI \-l " line\-length"
Sets the maximum length of lines.
.B hxnormalize
will wrap lines so that all lines are as long as possible, but no
longer than this length. Default is 72. Words that are longer than the
line length will not be broken, and will extend past this length. A
\(oqword\(cq is a sequence of characters delimited by white space.) The
content of the STYLE, SCRIPT and PRE elements will not be
line-wrapped.
.TP
.B \-s
Omit <span> tags that don't have any attributes.
.TP
.B \-L
Remove redundant \(oqlang\(cq and \(oqxml:lang\(cq attributes. (I.e.,
those whose value is the same as the language inherited from the
parent element.)
.TP
.BI \-c " commentmagic"
Comments are normally placed right after the preceding text. That is
usually correct for short comments, but some comments are meant to be
on a separate line.
.I commentmagic
is a string and when that string occurs inside a comment,
.B hxnormalize
will output an empty line before that comment. E.g. \fB\-c "===="\fR
can be used to put all comments that contain \(oq====\(cq on a separate
line, preceded by an empty line. By default, no comments are treated
that way.
.SH OPERANDS
The following operand is supported:
.TP 10
.I file-or-URL
The name or URL of an HTML file. If absent, standard input is read
instead.
.SH "EXIT STATUS"
The following exit values are returned:
.TP 10
.B 0
Successful completion.
.TP
.B > 0
An error occurred in the parsing of the HTML file.
.B hxnormalize
will try to correct the error and produce output anyway.
.SH ENVIRONMENT
To use a proxy to retrieve remote files, set the environment variables
.B http_proxy
and
.BR ftp_proxy "."
E.g.,
.B http_proxy="http://localhost:8080/"
.SH BUGS
.LP
The error recovery for incorrect HTML is primitive.
.LP
.B hxnormalize
will not omit an endtag if the white space after it could possibly be
significant. E.g., it will not remove the first </p> from
\(oq<div><p>text</p> <p>text</p></div>\(cq.
.LP
.B hxnormalize
can currently only retrieve remote files over HTTP. It doesn't handle
password-protected files, nor files whose content depends on HTTP
\(oqcookies.\(cq
.LP
When converting from XML to HTML (option
.B \-X
without option
.BR \-x ),
any pairs of \(OQ<![CDATA[\(CQ and \(oq]]>\(cq are removed and
character entities &lt; &gt; &quot; &apos; and &amp; are expanded (to
\(oq<\(cq, \(oq>\(cq, \(oq"\(cq, \(oq'\(cq and \(oq&\(cq,
respectively), but any other character entities are not expanded. To
expand other character entities, pipe the input through
.BR hxunent (1)
first.
.LP
To limit lines to a given number of characters,
.B hxnormalize
breaks lines at spaces (or inside tags). Some writing systems do not
use spaces between words and thus
.B hxnormalize
may not be able to break lines, except at already existing line
breaks.
.LP
To make short lines longer,
.B hxnormalize
will combine lines and replace a line break by a space, except in
writing systems that do not put spaces between words, where the line
break is replaced by nothing.
.B hxnormalize
currently only does the latter for Japanese, Chinese, Korean,
Khmer and Thai. (The text must be correctly marked up with
\(oqlang\(cq or \(oqxml:lang\(cq.)
.SH "SEE ALSO"
.BR asc2xml (1),
.BR xml2asc (1),
.BR hxunent (1),
.BR UTF-8 " (RFC 2279)"


ChangeLog
2019-10-05  Bert Bos  <bert@w3.org>

	* Published version 7.8.

	* textwrap.c, langinfo.c, hxnormalize.c: Added knowledge about
	languages that do not use spaces between words. In such languages,
	a newline should not be converted to a space in outc() in
	textwrap.c, but only to a break opportunity.

2019-08-28  Bert Bos  <bert@w3.org>

	* hxtoc.c: The element to group headings in HTML5 is called
	HGROUP, not HEADER. The heading of a section (SECTION, ARTICLE,
	etc.) need not be the first element, there may be non-header
	elements before it.

	* hxwls.c: Print "longdesc", "classid" or "codebase" in the second
	column for the corresponding attribute. Also recognize srcset
	(somewhat).

2018-08-03  Bert Bos  <bert@w3.org>

	* hxnormalize.c: Added option -X to indicate the input is XML
	instead of HTML. Handle conversion of CDATA elements to XML by
	escaping < and & instead of adding <![CDATA[. Added corresponding
	test normalize13.sh.

2018-04-29  Bert Bos  <bert@w3.org>

	* Released version 7.7.

	* dtd.hash: Don't include the arguments in the forward declaration
	of lookup_element(), because those arguments differ slightly
	depending on which version of gperf is used to generate dtd.c:
	"unsigned int" in gperf 3.0 vs "size_t" in gperf 3.1.

	* conficgure.ac: Check for libiconv with the AM_ICONV macro (from
	gettext) instead of AC_SEARCH_LIBS. (Advice from Christian
	Weisgerber, received via Anthony Bentley). That in turn requires
	adding config.rpath to the EXTRA_DIST in Makefile.am.

2018-02-23  Bert Bos  <bert@w3.org>

	* tests/index10.sh: Added a test for a term database.

	* hxindex.c: Replaced fgets() by getline(), to allow lines of
	arbitrary length in the term database.

2018-02-15  Bert Bos  <bert@w3.org>

	* tests/index9.sh: Added a test for hxindex's option -N.

	* tests/index5.sh: Updated for hxindex's option -N.

	* tests/cite5.sh: Added the example from hxmkbib.1 as a test.

	* tests/normalize12.sh: Added a test for the NOSCRIPT element.

	* hxindex.1: Added option -N.

	* hxmkbib.1: The example was missing some "%" signs.

	* hxmkdir.c: Read the template before reading the auxfile, so that
	a pipe like "hxcite -a auxfile bib template | hxmkbib -a auxfile
	bib" becomes possible. hxcite will have the time to finishe
	writing the auxfile before hxmkbib reads it.

2018-02-12  Bert Bos  <bert@w3.org>

	* hxindex.c: Added option -N.

	* dtd.hash: Since HTML5, the NOSCRIPT element is phrase-level
	instead of block-level.

2018-02-09  Bert Bos  <bert@w3.org>

	* Version 7.6 released.

	* selmatch.c: matches_sel() could read uninitialized memory
	(h->parent where h is the root of a tree) potentially leading to a
	SIGBUS. Thanks to Anthony J. Bentley for finding this bug.

2018-01-20  Bert Bos  <bert@w3.org>

	* Version 7.5 released.

	* hxcite.1: Spelling error.

	* hxincl.1: Spelling error.

	* url.c: Fixed unsafe way to call warnx(). Thanks to Robin
	Naundorf for these three corrections.

2017-12-08  Bert Bos  <bert@w3.org>

	* tests/normalize11.sh: Added to test for "main".

	* dtd.hash: Added the HTML5 element "main".

2017-12-07  Bert Bos  <bert@w3.org>

	* hxwls.c: Added support for HTML5 elements (source, audio, video)
	and for the classid and codebase attribute. Fixed an error: input
	has a src attribute, not href.

2017-11-24  Bert Bos  <bert@w3.org>

	* Version 7.4 released.

	* hxremove.1, hxselect.1: Updated.

	* tests/remove4.sh, tests/select18.sh, tests/select19.sh,
	tests/select20.sh, tests/select21.sh, tests/select22.sh,
	tests/remove5.sh tests/remove6.sh tests/remove7.sh,
	tests/select23.sh tests/select24.sh: Added.

	* selmatch.c: Now handles the ':not()' pseudo-class and
	comma-separated selectors. Fixed a bug in matching the ':root'
	pseudo-class.

	* hxremove.c, hxselect.c: parse_selector() has different
	arguments.

	* selector.c: Now parses the ':not()' pseudo-class and
	comma-separated selectors. Errors are now printed with a newline
	at the end. Added new functions to dump (serialize) a selector to
	a file, useful for debugging.

	* openurl.c: Functions now either return CURLcode or CURLMcode,
	but not some mixture.

2017-11-11  Bert Bos  <bert@w3.org>

	* dtd.hash: Added some HTML5 elements: source, track, wbr (because
	they are empty) and audio and video (as parents for source and
	track).

	* toc4.sh, normalize10.sh: Added tests for the <source> element.

	* Released version 7.3.

2017-10-16  Bert Bos  <bert@w3.org>

	* normalize9.sh: Added test for A inside FIGCAPTION.

	* dtd.hash: An A can now be inside a FIGCAPTION. Made source
	shorter and more readable with some macros.

2017-10-06  Bert Bos  <bert@w3.org>

	* Released version 7.2.

	* tests/extract1.sh: Added option -N to nc, to close the
	connection after EOF on stdin, because hxextract (with libcurl)
	seems to wait for the closing of the connection.

	* tests/unent1.ssh, tests/unent2.ssh, tests/unent3.ssh,
	tests/unent4.ssh, tests/unent5.ssh: Added tests for hxunent.

	* hxunentmain.c: Fixed bug reported by Laurent Martelli: When a
	line was longer than the 4096-byte buffer, an entity could be on
	the edge between two buffers and not be recognized.

2017-08-13  Bert Bos  <bert@w3.org>

	* extract1.sh: return code 77 (skip) rather than 0 when netcat is
	unavailable.

	* tests/pipe4.sh, tests/select01.sh, ..., tests/select17.sh,
	tests/unpipe6.sh: New tests.

	* scan.l: Accept non-ASCII characters in names (element names,
	attribute names, etc.)

	* hxselect.c, hxremove.c: Put their common code in
	selmatch.c. hxselect now uses the data structures from tree.c,
	just like hxremove.c. The non-element selector '::attr(...)' is
	now also supported.

	* selector.c, selmatch.c: Added support for more structural
	pseudo-classes (last-child, nth-last-of-type, etc.). Generalized
	support for backslash escapes. Support hexadecimal backslash
	escapes also outside ASCII range (such characters are converted to
	UTF-8). Bug fixes.

2016-10-17  Bert Bos  <bert@w3.org>

	* Released version 7.1.


hxnormalize.c
/*
 * Format an HTML source in a consistent manner.
 *
 * Copyright © 1994-2012 World Wide Web Consortium
 * See http://www.w3.org/Consortium/Legal/2002/copyright-software-20021231
 *
 * Created 9 May 1998
 * Bert Bos <bert@w3.org>
 * $Id: hxnormalize.c,v 1.25 2019/10/05 17:11:31 bbos Exp $
 */
#include "config.h"
#include <stdio.h>
#include <ctype.h>
#include <stdlib.h>
#ifdef HAVE_UNISTD_H
#  include <unistd.h>
#endif
#ifdef HAVE_STRING_H
#  include <string.h>
#elif HAVE_STRINGS_H
#  include <strings.h>
#endif
#include <assert.h>
#include <stdbool.h>
#include "export.h"
#include "types.e"
#include "tree.e"
#include "html.e"
#include "scan.e"
#include "textwrap.e"
#include "dict.e"
#include "openurl.e"
#include "errexit.e"
#include "langinfo.e"

static Tree tree;
static bool do_xml = false;
static bool do_endtag = false;
static bool has_errors = false;
static bool do_doctype = true;
static bool clean_span = false;
static string long_comment = NULL;
static bool do_lang = false;
static bool input_is_xml = false;


/* handle_error -- called when a parse error occurred */
void handle_error(void *clientdata, const string s, int lineno)
{
  fprintf(stderr, "%d: %s\n", lineno, s);
  has_errors = true;
}

/* start -- called before the first event is reported */
void* start(void)
{
  tree = create();
  return NULL;
}
  
/* end -- called after the last event is reported */
void end(void *clientdata)
{
  /* skip */
}

/* handle_comment -- called after a comment is parsed */
void handle_comment(void *clientdata, string commenttext)
{
  tree = append_comment(tree, commenttext);
}

/* handle_html_text -- called after a text chunk is parsed */
void handle_html_text(void *clientdata, string text)
{
  tree = append_text(tree, text);
}

/* handle_text -- called after a text chunk is parsed */
void handle_text(void *clientdata, string text)
{
  tree = tree_append_text(tree, text);
}

/* handle_decl -- called after a declaration is parsed */
void handle_decl(void *clientdata, string gi,
		 string fpi, string url)
{
  tree = append_declaration(tree, gi, fpi, url);
}

/* handle_pi -- called after a PI is parsed */
void handle_pi(void *clientdata, string pi_text)
{
  tree = append_procins(tree, pi_text);
}

/* handle_html_starttag -- called after a start tag is parsed */
void handle_html_starttag(void *clientdata, string name, pairlist attribs)
{
  tree = html_push(tree, name, attribs);
  free(name);
}

/* handle_starttag -- called after a start tag is parsed */
void handle_starttag(void *clientdata, string name, pairlist attribs)
{
  tree = tree_push(tree, name, attribs);
  free(name);
}

/* handle_html_emptytag -- called after an empty tag is parsed */
void handle_html_emptytag(void *clientdata, string name, pairlist attribs)
{
  tree = html_push(tree, name, attribs);
  free(name);
}

/* handle_emptytag -- called after an empty tag is parsed */
void handle_emptytag(void *clientdata, string name, pairlist attribs)
{
  tree = tree_push(tree, name, attribs);
  tree = tree_pop(tree, name);
  free(name);
}

/* handle_html_endtag -- called after an endtag is parsed (name may be "") */
void handle_html_endtag(void *clientdata, string name)
{
  tree = html_pop(tree, name);
  free(name);
}

/* handle_endtag -- called after an endtag is parsed (name may be "") */
void handle_endtag(void *clientdata, string name)
{
  tree = tree_pop(tree, name);
  free(name);
}

/* insert -- insert an attribute into a sorted list of attributes */
static pairlist insert(pairlist x, pairlist list)
{
  if (! list) {					/* Empty list */
    x->next = NULL;
    return x;
  } else if (strcmp(x->name, list->name) <= 0) { /* Insert at head */
    x->next = list;
    return x;
  } else {					/* Insert not at head */
    list->next = insert(x, list->next);
    return list;
  }
}

/* sort_list -- sort a linked list of attributes, return reordered list */
static pairlist sort_list(pairlist list)
{
  /* Insertion sort should be fast enough... */
  if (! list) return NULL;
  else return insert(list, sort_list(list->next));
}

/* next_ambiguous -- check if omitting end changes the meaning */
static bool next_ambiguous(Node *n)
{
  Node *h = n;

  /* Skip text nodes with only white space */
  while (h->sister && h->sister->tp == Text && only_space(h->sister->text))
    h = h->sister;

  if (h->sister == NULL) return false;
  if (h->sister->tp == Text) return true;
  if (h->sister->tp == Comment) return true;
  if (h->sister->tp == Procins) return true;
  if (h->sister->tp == Declaration) return false; /* Should not occur */
  assert(h->sister->tp == Element);		/* Cannot be Root */
  return has_parent(h->sister->name, n->name);
}

/* needs_quotes -- check if the attribute value can be printed unquoted */
static bool needs_quotes(const string s)
{
  int i;
  assert(s);
  if (!s[0]) return true;			/* Empty string */
  for (i = 0; s[i]; i++)
    if (!isalnum(s[i]) && (s[i] != '-') && (s[i] != '.')) return true;
  return false;
}

/* pp -- print the document normalized */
static void pp(Tree n, bool preformatted, bool allow_text,
	       conststring lang)
{
  bool pre, mixed, with_space;
  conststring lang2;
  string s;
  pairlist h;
  size_t i, j;
  Tree l;

  with_space = with_spaces(lang);	/* Language with spaces between words? */

  switch (n->tp) {
    case Text:
      if (!allow_text) {
	assert(only_space(n->text));
      } else {
	s = n->text;
	out(s, preformatted, with_space);
      }
      break;
    case Comment:
      if (long_comment && strstr(n->text, long_comment) && !preformatted) {
	/* Found a comment that should have an empty line before it */
	outbreak();
	outln(NULL, true, with_space);
      }
      out("<!--", true, true); out(n->text, true, with_space);
      if (allow_text || preformatted) out("-->", true, true);
      else outln("-->", preformatted, true);
      break;
    case Declaration:
      if (do_doctype) {
	out("<!DOCTYPE ", false, true);
	out(n->name, false, true);
	if (n->text) {
	  out(" PUBLIC \"", false, true);
	  out(n->text, false, true);
	  out("\"", false, true);
	}
	if (n->url) {
	  if (!n->text) out(" SYSTEM", false, true);
	  out(" \"", false, true);
	  out(n->url, false, true);
	  out("\"", false, true);
	} else if (n->text && do_xml) {	/* XML cannot omit the system literal */
	  out(" \"\"", false, true);
	}
	outln(">", false, true);
      }
      break;
    case Procins:
      out("<?", false, true); out(n->text, true, true);
      if (allow_text || preformatted) out(">", false, true);
      else outln(">", false, true);
      break;
    case Element:
      if (clean_span && eq(n->name, "span") && ! n->attribs) {
	/* Omit start and end tags, print just the children. */
	for (l = n->children; l != NULL; l = l->sister)
	  pp(l, preformatted, true, lang);
	break;
      }
      /* Check for language attribute. */
      lang2 = pairlist_get(n->attribs, "lang");
      if (!lang2) lang2 = pairlist_get(n->attribs, "xml:lang");

      /* Remove redundant language attribute */
      if (do_lang && lang && lang2 && eq(lang, lang2)) {
	pairlist_unset(&n->attribs, "lang");
	pairlist_unset(&n->attribs, "xml:lang");
      }

      /* Update inherited language. */
      if (lang2) lang = lang2;

      if (!preformatted && break_before(n->name)) outln(NULL, false, true);
      out("<", preformatted, true); out(n->name, preformatted, true);
      if (break_before(n->name)) inc_indent();
      n->attribs = sort_list(n->attribs);
      for (h = n->attribs; h != NULL; h = h->next) {
	out(" ", false, true); out(h->name, false, true);
	if (do_xml) {
	  out("=\"", false, true);
	  out(h->value ? h->value : h->name, true, true);
	  outc('"', false, true);
	} else if (h->value == NULL) {
	  /* The h->name *is* the value (and the attribute name is implicit) */
	} else if (!needs_quotes(h->value)) {
	  out("=", false, true); /* Omit the quotes */
	  out(h->value, true, true);
	} else {
	  out("=\"", false, true);
	  out(h->value, true, true);
	  outc('"', false, true);
	}
      }
      if (is_empty(n->name)) {
	assert(n->children == NULL);
	outbreakpoint();
	out(do_xml ? " />" : ">", true, true);
	if (break_before(n->name)) dec_indent();
	if (!preformatted && break_after(n->name)) outln(NULL, false, true);

      } else if (do_xml && !input_is_xml && is_cdata_elt(n->name)) {
	/* Escape '<' and '&', but only if input was HTML, not XML */
	if (!n->children) {
	  out(" />", true, true);
	  if (break_before(n->name)) dec_indent();
	} else {
	  outbreakpoint();
	  out(">", preformatted, true);
	  for (l = n->children; l; l = l->sister) {
	    assert(l->tp == Text);
	    for (s = l->text; *s; s++)
	      if (*s == '<') out("&lt;", true, true);
	      else if (*s == '&') out("&amp;", true, true);
	      else outc(*s, true, with_space);
	  }
	  if (break_before(n->name)) dec_indent();
	  out("</", true, true);
	  out(n->name, true, true);
	  outbreakpoint();
	  out(">", preformatted, true);
	}
	if (!preformatted && break_after(n->name)) outbreak();

      } else if (!do_xml && input_is_xml && is_cdata_elt(n->name) &&
		 n->children) {
	/* Remove "<![CDATA[" and "]]>", or unescape &lt; and &amp;,
	   but only if input was XML, not HTML */
	outbreakpoint();
	out(">", preformatted, true);
	for (l = n->children; l != NULL; l = l->sister) {
	  if (l->tp != Text) {
	    errexit("Cannot convert <%s> to HTML because it has children\n",
		    n->name);
	  } else if (hasprefix(l->text, "<![CDATA[")) {
	    assert(hasaffix(l->text, "]]>"));
	    s = l->text + 9;	/* Skip "<![CDATA[" */
	    i = strlen(s) - 3;	/* Omit "]]>" */
	    for (j = 0; j < i; j++) outc(s[j], true, with_space);
	  } else {		/* Unescape &lt; and &amp; */
	    for (s = l->text; *s; s++)
	      if (hasprefix(s, "&amp;")) {outc('&', true, true); s += 4;}
	      else if (hasprefix(s, "&lt;")) {outc('<', true, true); s += 3;}
	      else if (hasprefix(s, "&gt;")) {outc('>', true, true); s += 3;}
	      else if (hasprefix(s, "&quot;")) {outc('"', true, true); s += 5;}
	      else if (hasprefix(s, "&apos;")) {outc('\'', true, true); s += 5;}
	      else outc(*s, true, with_space);
	  }
	}
	if (break_before(n->name)) dec_indent();
	out("</", preformatted, true);
	out(n->name, preformatted, true);
	outbreakpoint();
	out(">", preformatted, true);
	if (!preformatted && break_after(n->name)) outbreak();

      } else {
	outbreakpoint();
	out(">", preformatted, true);
	pre = preformatted || is_pre(n->name);
	mixed = is_mixed(n->name);
	for (l = n->children; l != NULL; l = l->sister)
	  pp(l, pre, mixed, lang);
	if (break_before(n->name)) dec_indent();
	if (do_xml || do_endtag || need_etag(n->name) || next_ambiguous(n)) {
	  out("</", pre, true); out(n->name, pre, true);
	  outbreakpoint();
	  out(">", preformatted, true);
	}
	if (!preformatted && break_after(n->name)) outbreak();
      }
      break;
    default:
      assert(!"Cannot happen");
  }
}

/* prettyprint -- print the tree normalized */
static void prettyprint(Tree t)
{
  Tree h;
  assert(t->tp == Root);
  for (h = t->children; h != NULL; h = h->sister) pp(h, false, false, NULL);
  flush();
}

/* usage -- print usage message and exit */
static void usage(string prog)
{
  fprintf(stderr, "%s version %s\n\
Usage: %s [-e] [-d] [-x] [-L] [-i indent] [-l linelen] [-c commentmagic] [file_or_url]\n",
	  prog, VERSION, prog);
  exit(1);
}

/* main -- main body */
int main(int argc, char *argv[])
{
  int c, status = 200;

  while ((c = getopt(argc, argv, "edxXi:l:sc:L")) != -1)
    switch (c) {
    case 'e': do_endtag = true; break;
    case 'x': do_xml = true; break;
    case 'X': input_is_xml = true; break;
    case 'd': do_doctype = false; break;
    case 'i': set_indent(atoi(optarg)); break;
    case 'l': set_linelen(atoi(optarg)); break;
    case 's': clean_span = true; break;
    case 'c': long_comment = optarg; break;
    case 'L': do_lang = true; break;
    default: usage(argv[0]);
    }
  if (optind == argc) yyin = stdin;
  else if (optind == argc - 1) yyin = fopenurl(argv[optind], "r", &status);
  else usage(argv[0]);
  if (yyin == NULL) {perror(argv[optind]); exit(2);}
  if (status != 200) errexit("%s : %s\n", argv[optind], http_strerror(status));

  /* Bind the parser callback routines to our handlers */
  if (input_is_xml) {
    set_error_handler(handle_error);
    set_start_handler(start);
    set_end_handler(end);
    set_comment_handler(handle_comment);
    set_text_handler(handle_text);
    set_decl_handler(handle_decl);
    set_pi_handler(handle_pi);
    set_starttag_handler(handle_starttag);
    set_emptytag_handler(handle_emptytag);
    set_endtag_handler(handle_endtag);
  } else {			/* Input is HTML */
    set_error_handler(handle_error);
    set_start_handler(start);
    set_end_handler(end);
    set_comment_handler(handle_comment);
    set_text_handler(handle_html_text);
    set_decl_handler(handle_decl);
    set_pi_handler(handle_pi);
    set_starttag_handler(handle_html_starttag);
    set_emptytag_handler(handle_html_emptytag);
    set_endtag_handler(handle_html_endtag);
  }

  if (yyparse() != 0) {exit(3);}
  tree = get_root(tree);
  prettyprint(tree);
  return has_errors ? 1 : 0;
}
